src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/TraceEventExtensions.cs

```diff
@@ -57,26 +58,274 @@ public static bool TryGetCounterPayload(this TraceEvent traceEvent, CounterFilte
+                payload = new GaugePayload(meterName, instrumentName, null, unit, tags, lastValue, obj.TimeStamp);
+            }
+            else
+            {
+                // for observable instruments we assume the lack of data is meaningful and remove it from the UI
+                // this happens when the Gauge callback function throws an exception.
+
+                //TODO Can this occur for other meter types?
+                payload = new CounterEndedPayload(meterName, instrumentName, null, obj.TimeStamp);
+                
+            }
+        }
+
+        private static void HandleCounterRate(TraceEvent traceEvent, CounterFilter filter, string sessionId, out ICounterPayload payload)
+        {
+            payload = null;
+
+            string payloadSessionId = (string)traceEvent.PayloadValue(0);
+
+            if (payloadSessionId != sessionId)
+            {
+                return;
+            }
+
+            string meterName = (string)traceEvent.PayloadValue(1);
+            //string meterVersion = (string)obj.PayloadValue(2);
+            string instrumentName = (string)traceEvent.PayloadValue(3);
+            string unit = (string)traceEvent.PayloadValue(4);
+            string tags = (string)traceEvent.PayloadValue(5);
+            string rateText = (string)traceEvent.PayloadValue(6);
+
+            if (!filter.IsIncluded(meterName, instrumentName))
+            {
+                return;
+            }
+
+            if (double.TryParse(rateText, out double rate))
+            {
+                payload = new RatePayload(meterName, instrumentName, null, unit, tags, rate, filter.IntervalSeconds, traceEvent.TimeStamp);
+            }
+        }
+
+        private static void HandleHistogram(TraceEvent obj, CounterFilter filter, string sessionId, out List<ICounterPayload> payload)
+        {
+            payload = new List<ICounterPayload>();
+
+            string payloadSessionId = (string)obj.PayloadValue(0);
+
+            if (payloadSessionId != sessionId)
+            {
+                return;
+            }
+
+            string meterName = (string)obj.PayloadValue(1);
+            //string meterVersion = (string)obj.PayloadValue(2);
+            string instrumentName = (string)obj.PayloadValue(3);
+            string unit = (string)obj.PayloadValue(4);
+            string tags = (string)obj.PayloadValue(5);
+            string quantilesText = (string)obj.PayloadValue(6);
+
+            if (!filter.IsIncluded(meterName, instrumentName))
+            {
+                return;
+            }
+
+            KeyValuePair<double, double>[] quantiles = ParseQuantiles(quantilesText);
+            foreach ((double key, double val) in quantiles)
+            {
+                string tagsWithPercentile = AppendPercentile(tags, FormattableString.Invariant($"Percentile={(int)(100*key)}"));
+                payload.Add(new PercentilePayload(meterName, instrumentName, null, unit, tagsWithPercentile, val, obj.TimeStamp));
```

Comment:
Per Justin: Consider merging PercentilePayloads into 1 HistogramPayload.

Reply:
I like this idea. I think its reasonable to assume that the data you get in this event will be always be a set of three percentiles. There is no need to introduce complexity downstream handling an arbitrary number of percentiles. If in the future we did add an alternate event representation we can pay the complexity cost then and tailor the new code to whatever new representation we chose.